Fix BDC remember password and reprompting connection

Charles-Gagnon opened the pull request:
Fixes #7921

Fixed a couple of issues.

First off was the issue that even if you left remember password unchecked the node would still "remember" it and thus not prompt you for the password when you opened the dashboard. But if you restarted ADS then it would correctly not store it and so would fail to open the dashboard.

I've changed it to not save the password to the credential store, and then to re-prompt for the password if it doesn't exist (or the stored password is wrong, such as if it changed). The password will persist for the session though so the user won't be prompted unless an error occurs. 

Other fixes :

- Fixed error message displayed on Connection dialog to be more friendly (before it mentioned something about endpoints which isn't helpful in that scenario)
- Made dashboard open when new controller connection is created
- Added error handling for model
- Added reprompt connection for model if there's no valid controller (for example if the user cancelled out of the connection dialog when the dashboard was opened)

Follow-up items not included in this PR :
- Still need to make the dashboard indicate an error occurred (right now it just sits there spinning)

Charles-Gagnon commented:
@kevcunnane @alanrenmsft Could you take a look at this when you get a chance?

kevcunnane commented on extensions/big-data-cluster/src/bigDataCluster/dialog/hdfsDialogBase.ts:
```diff
@@ -28,6 +28,12 @@ export interface HdfsDialogProperties {
 }
 
+export class HdfsDialogCancelledError extends Error {
+	constructor(message: string = 'Dialog cancelled') {
```
Does this need localization?

Charles-Gagnon replied:
Depends. I generally go the route of "error messages shouldn't be exposed to the user and so don't need to be localized". But I also know we don't do that everywhere so if we think it's appropriate to default to localized errors I can switch it. I just don't want to make debugging a pain by having localized error messages in the console logs.

alanrenmsft commented on extensions/big-data-cluster/src/bigDataCluster/dialog/bdcDashboardModel.ts:
```diff
@@ -21,12 +22,22 @@ export class BdcDashboardModel {
+		try {
+			this._clusterController = new ClusterController(options.url, options.auth, options.username, options.password, ignoreSslVerification);
+			this.refresh();
+		} catch {
```
what error are you trying to catch here, the new ClusterController()? this.refresh is async and the error won't be caught here

Charles-Gagnon replied:
ClusterController throws an error in the constructor if the username or password is empty and basic auth is selected

alanrenmsft commented on extensions/big-data-cluster/src/bigDataCluster/dialog/bdcDashboardModel.ts:
```diff
@@ -46,21 +57,28 @@ export class BdcDashboardModel {
+		try {
+			if (!this._clusterController) {
+				// If this succeeds without error we know we have a clusterController at this point
+				await this.promptReconnect();
```
what happens if user closes the connect dialog, it seems we are handling that scenario here.

Charles-Gagnon replied:
The dashboard is notified that an error has occurred but right now we don't handle that in any special way so nothing happens (it sits there spinning). The user can press refresh to cause the dialog to appear again and if the correct information is entered then the dashboard will refresh with new information as expected.

I have a follow up task to make the dashboard indicate if an error occurred but that won't be a part of this PR.

Charles-Gagnon commented:
@kevcunnane I changed it as discussed earlier so it now keeps the password for the session (re-prompting if the credentials are wrong though). 

I also fixed up some no-floating-promises errors
